## Data Loading and Validation

Goals:

-Can we load the data?

-What does it contain?

-What is missing?

-What time period does it cover?

-Are the dtypes correct?

-What will our prediction target eventually be?

### Project goal: Predict household global active power consumption 60 minutes into the future using historical electrical measurements, lagged consumption, and calendar features.

source found for dataset:

https://archive.ics.uci.edu/dataset/235/individual%2Bhousehold%2Belectric%2Bpower%2Bci

In [1]:
#check proper tools are installed including the uci helper package

#conda activate pytorch
#python -m pip install ucimlrepo pandas pyarrow scikit-learn matplotlib

In [2]:
from ucimlrepo import fetch_ucirepo

power = fetch_ucirepo(id=235)

print(power.metadata)

{'uci_id': 235, 'name': 'Individual Household Electric Power Consumption', 'repository_url': 'https://archive.ics.uci.edu/dataset/235/individual+household+electric+power+consumption', 'data_url': 'https://archive.ics.uci.edu/static/public/235/data.csv', 'abstract': 'Measurements of electric power consumption in one household with a one-minute sampling rate over a period of almost 4 years. Different electrical quantities and some sub-metering values are available.', 'area': 'Physics and Chemistry', 'tasks': ['Regression', 'Clustering'], 'characteristics': ['Multivariate', 'Time-Series'], 'num_instances': 2075259, 'num_features': 9, 'feature_types': ['Real'], 'demographics': [], 'target_col': None, 'index_col': None, 'has_missing_values': 'no', 'missing_values_symbol': None, 'year_of_dataset_creation': 2006, 'last_updated': 'Fri Mar 08 2024', 'dataset_doi': '10.24432/C58K54', 'creators': ['Georges Hebrail', 'Alice Berard'], 'intro_paper': None, 'additional_info': {'summary': 'This archiv

/home/dmdgxs/miniforge3/envs/pytorch/lib/python3.12/site-packages/ucimlrepo/fetch.py:97: DtypeWarning: Columns (2: Global_active_power, 3: Global_reactive_power, 4: Voltage, 5: Global_intensity, 6: Sub_metering_1, 7: Sub_metering_2) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(data_url)


In [3]:
X = power.data.features

print(type(X))
print(X.shape)
print(X.head())

<class 'pandas.DataFrame'>
(2075259, 9)
         Date      Time Global_active_power Global_reactive_power  Voltage  \
0  16/12/2006  17:24:00               4.216                 0.418  234.840   
1  16/12/2006  17:25:00               5.360                 0.436  233.630   
2  16/12/2006  17:26:00               5.374                 0.498  233.290   
3  16/12/2006  17:27:00               5.388                 0.502  233.740   
4  16/12/2006  17:28:00               3.666                 0.528  235.680   

  Global_intensity Sub_metering_1 Sub_metering_2  Sub_metering_3  
0           18.400          0.000          1.000            17.0  
1           23.000          0.000          1.000            16.0  
2           23.000          0.000          2.000            17.0  
3           23.000          0.000          1.000            17.0  
4           15.800          0.000          1.000            17.0  


In [4]:
print(X.columns.tolist())

['Date', 'Time', 'Global_active_power', 'Global_reactive_power', 'Voltage', 'Global_intensity', 'Sub_metering_1', 'Sub_metering_2', 'Sub_metering_3']


In [5]:
X.info()

<class 'pandas.DataFrame'>
RangeIndex: 2075259 entries, 0 to 2075258
Data columns (total 9 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   Date                   str    
 1   Time                   str    
 2   Global_active_power    object 
 3   Global_reactive_power  object 
 4   Voltage                object 
 5   Global_intensity       object 
 6   Sub_metering_1         object 
 7   Sub_metering_2         object 
 8   Sub_metering_3         float64
dtypes: float64(1), object(6), str(2)
memory usage: 176.0+ MB


In [6]:
X.head(10)

,Date,Time,Global_active_power,Global_reactive_power,Voltage,Global_intensity,Sub_metering_1,Sub_metering_2,Sub_metering_3
0,16/12/2006,17:24:00,4.216,0.418,234.840,18.400,0.000,1.000,17.0
1,16/12/2006,17:25:00,5.360,0.436,233.630,23.000,0.000,1.000,16.0
2,16/12/2006,17:26:00,5.374,0.498,233.290,23.000,0.000,2.000,17.0
3,16/12/2006,17:27:00,5.388,0.502,233.740,23.000,0.000,1.000,17.0
4,16/12/2006,17:28:00,3.666,0.528,235.680,15.800,0.000,1.000,17.0
5,16/12/2006,17:29:00,3.520,0.522,235.020,15.000,0.000,2.000,17.0
6,16/12/2006,17:30:00,3.702,0.520,235.090,15.800,0.000,1.000,17.0
7,16/12/2006,17:31:00,3.700,0.520,235.220,15.800,0.000,1.000,17.0
8,16/12/2006,17:32:00,3.668,0.510,233.990,15.800,0.000,1.000,17.0
9,16/12/2006,17:33:00,3.662,0.510,233.860,15.800,0.000,2.000,16.0


In [7]:
# missing values check

missing = X.isna().sum()

print(missing)

Date                         0
Time                         0
Global_active_power          0
Global_reactive_power        0
Voltage                      0
Global_intensity             0
Sub_metering_1               0
Sub_metering_2               0
Sub_metering_3           25979
dtype: int64


In [8]:
missing_pct = (
    X.isna().mean() * 100
).sort_values(ascending=False)

print(missing_pct)

Sub_metering_3           1.251844
Date                     0.000000
Time                     0.000000
Global_active_power      0.000000
Global_reactive_power    0.000000
Voltage                  0.000000
Global_intensity         0.000000
Sub_metering_1           0.000000
Sub_metering_2           0.000000
dtype: float64


In [9]:
# time range inspection

print(X.head())
print(X.dtypes)

         Date      Time Global_active_power Global_reactive_power  Voltage  \
0  16/12/2006  17:24:00               4.216                 0.418  234.840   
1  16/12/2006  17:25:00               5.360                 0.436  233.630   
2  16/12/2006  17:26:00               5.374                 0.498  233.290   
3  16/12/2006  17:27:00               5.388                 0.502  233.740   
4  16/12/2006  17:28:00               3.666                 0.528  235.680   

  Global_intensity Sub_metering_1 Sub_metering_2  Sub_metering_3  
0           18.400          0.000          1.000            17.0  
1           23.000          0.000          1.000            16.0  
2           23.000          0.000          2.000            17.0  
3           23.000          0.000          1.000            17.0  
4           15.800          0.000          1.000            17.0  
Date                         str
Time                         str
Global_active_power       object
Global_reactive_power     obje

In [10]:
# combine date and time into one feature

import pandas as pd

X["datetime"] = pd.to_datetime(
    X["Date"] + " " + X["Time"],
    dayfirst=True
)

In [11]:
print("Start:", X["datetime"].min())
print("End:", X["datetime"].max())

Start: 2006-12-16 17:24:00
End: 2010-11-26 21:02:00


In [12]:
# convert measurement columns to numeric

numeric_cols = [
    "Global_active_power",
    "Global_reactive_power",
    "Voltage",
    "Global_intensity",
    "Sub_metering_1",
    "Sub_metering_2",
    "Sub_metering_3"
]

for col in numeric_cols:
    X[col] = pd.to_numeric(
        X[col],
        errors="coerce" # malformed strings will be converted to NaN
    )

In [13]:
X.dtypes

Date                                str
Time                                str
Global_active_power             float64
Global_reactive_power           float64
Voltage                         float64
Global_intensity                float64
Sub_metering_1                  float64
Sub_metering_2                  float64
Sub_metering_3                  float64
datetime                 datetime64[us]
dtype: object

In [14]:
X[numeric_cols].isna().sum()

Global_active_power      25979
Global_reactive_power    25979
Voltage                  25979
Global_intensity         25979
Sub_metering_1           25979
Sub_metering_2           25979
Sub_metering_3           25979
dtype: int64

In [15]:
X = X.sort_values("datetime").reset_index(drop=True)

In [16]:
print("Rows:", len(X))
print("Start:", X["datetime"].min())
print("End:", X["datetime"].max())

Rows: 2075259
Start: 2006-12-16 17:24:00
End: 2010-11-26 21:02:00


In [17]:
print(
    "Duplicate timestamps:",
    X["datetime"].duplicated().sum()
)

Duplicate timestamps: 0


In [18]:
# add missing timestamps

full_range = pd.date_range(
    start=X["datetime"].min(),
    end=X["datetime"].max(),
    freq="1min"
)

missing_timestamps = full_range.difference(
    X["datetime"]
)

print(
    "Missing timestamps:",
    len(missing_timestamps)
)

Missing timestamps: 0


In [19]:
# check if missing values occur together

missing_rows = X[
    X[numeric_cols].isna().any(axis=1)
]

print(missing_rows.shape)
missing_rows.head(20)

(25979, 10)


,Date,Time,Global_active_power,Global_reactive_power,Voltage,Global_intensity,Sub_metering_1,Sub_metering_2,Sub_metering_3,datetime
6839,21/12/2006,11:23:00,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2006-12-21 11:23:00
6840,21/12/2006,11:24:00,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2006-12-21 11:24:00
19724,30/12/2006,10:08:00,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2006-12-30 10:08:00
19725,30/12/2006,10:09:00,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2006-12-30 10:09:00
41832,14/1/2007,18:36:00,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2007-01-14 18:36:00
61909,28/1/2007,17:13:00,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2007-01-28 17:13:00
98254,22/2/2007,22:58:00,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2007-02-22 22:58:00
98255,22/2/2007,22:59:00,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2007-02-22 22:59:00
142588,25/3/2007,17:52:00,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2007-03-25 17:52:00
190497,28/4/2007,00:21:00,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2007-04-28 00:21:00


In [20]:
# do the missing values occur in blocks?

missing_all = X[numeric_cols].isna().all(axis=1)

groups = (
    missing_all.ne(missing_all.shift())
    .cumsum()
)

missing_blocks = (
    X.loc[missing_all]
    .groupby(groups[missing_all])
    .agg(
        start=("datetime", "min"),
        end=("datetime", "max"),
        rows=("datetime", "size")
    )
)

missing_blocks.sort_values(
    "rows",
    ascending=False
).head(20)

,start,end,rows
138,2010-08-17 21:02:00,2010-08-22 21:27:00,7226
140,2010-09-25 03:56:00,2010-09-28 19:12:00,5237
14,2007-04-28 00:21:00,2007-04-30 14:23:00,3723
100,2009-06-13 00:30:00,2009-06-15 07:34:00,3305
118,2010-01-12 14:53:00,2010-01-14 19:01:00,3129
126,2010-03-20 03:52:00,2010-03-21 13:38:00,2027
104,2009-08-13 05:00:00,2009-08-13 19:50:00,891
32,2007-07-15 16:49:00,2007-07-15 18:11:00,83
78,2008-12-10 10:48:00,2008-12-10 11:57:00,70
34,2007-07-15 18:21:00,2007-07-15 19:07:00,47


In [21]:
# considering the large blocks of missing values
# Preserve the missing periods, 
# construct the forecasting target and lag features while the full timeline is intact, 
# and then remove rows that cannot produce a valid training example.

In [22]:
# full chronological timeline
#         ↓
# preserve NaNs/gaps
#         ↓
# create time features
#         ↓
# create lag features
#         ↓
# create 60-minute-ahead target
#         ↓
# identify valid examples
#         ↓
# drop unusable rows

In [23]:
# final check for missing or dupe timestamps

full_range = pd.date_range(
    start=X["datetime"].min(),
    end=X["datetime"].max(),
    freq="1min"
)

missing_timestamps = full_range.difference(
    X["datetime"]
)

print("Expected timestamps:", len(full_range))
print("Actual rows:", len(X))
print("Missing timestamps:", len(missing_timestamps))
print("Duplicate timestamps:", X["datetime"].duplicated().sum())

Expected timestamps: 2075259
Actual rows: 2075259
Missing timestamps: 0
Duplicate timestamps: 0


### Missing-data finding

All seven electrical measurement variables contain 25,979 missing
observations, and the missing values occur together across measurement
columns.

The missingness is concentrated in contiguous outages rather than
isolated cells. Several gaps span multiple days; the six largest blocks
account for roughly 95% of all missing measurements.

Because long-gap interpolation would fabricate substantial amounts of
electrical-consumption data, missing measurements will not be
interpolated. The complete timestamp sequence will instead be preserved
while lag variables and the forecasting target are constructed. Rows
without sufficient historical predictors or a valid future target will
then be excluded from model training.

In [24]:
# save to parquet

from pathlib import Path

processed_dir = Path("../data/processed")
processed_dir.mkdir(parents=True, exist_ok=True)

output_path = processed_dir / "household_power_validated.parquet"

X.to_parquet(
    output_path,
    index=False
)

print("Saved:", output_path)
print("Shape:", X.shape)

Saved: ../data/processed/household_power_validated.parquet
Shape: (2075259, 10)
